daemon/daemon.go

```diff
@@ -88,13 +92,69 @@ func (d *Daemon) GetProxy() *proxy.Proxy {
 }
 
 func (d *Daemon) WriteEndpoint(e *endpoint.Endpoint) error {
+	d.conf.LXCMapMU.Lock()
+	defer d.conf.LXCMapMU.Unlock()
 	if err := d.conf.LXCMap.WriteEndpoint(e); err != nil {
 		return fmt.Errorf("Unable to update eBPF map: %s", err)
 	}
 
 	return nil
 }
 
+// QueueEndpoint puts the given request in the endpoints queue for processing.
+// The given request will receive 'true' in the MyTurn channel whenever it's its
+// turn or false if the request was denied/canceled.
+func (d *Daemon) QueueEndpoint(req *endpoint.Request) {
+	go func(req *endpoint.Request) {
+		d.uniqueIDMU.Lock()
+		// We are skipping new requests, but only if the endpoint has not
+		// started its build process, since the endpoint is already in queue.
+		if isBuilding, exists := d.uniqueID[req.ID]; !isBuilding && exists {
+			req.MyTurn <- false
+		} else {
+			// We mark the request "not building" state and send it to
+			// the building queue.
+			d.uniqueID[req.ID] = false
+			d.buildEndpointChan <- req
+		}
+		d.uniqueIDMU.Unlock()
+	}(req)
+}
+
+// RemoveFromEndpointQueue removes the endpoint from the queue.
+func (d *Daemon) RemoveFromEndpointQueue(epID uint16) {
+	d.uniqueIDMU.Lock()
+	delete(d.uniqueID, epID)
+	d.uniqueIDMU.Unlock()
+}
+
+// EndpointBuilder creates #CPUs go routines that listen on the
+// `d.buildEndpointChan` for new endpoints.
+func (d *Daemon) EndpointBuilder() {
+	log.Debugf("Creating %d worker threads", runtime.NumCPU())
+	// Create the same amount of worker threads as there are CPUs
+	for w := 0; w < runtime.NumCPU(); w++ {
+		go func() {
+			for e := range d.buildEndpointChan {
+				d.uniqueIDMU.Lock()
+				// Set the endpoint to "building" state
+				d.uniqueID[e.ID] = true
+				e.MyTurn <- true
+				d.uniqueIDMU.Unlock()
+				// Wait for the endpoint to build
+				<-e.Done
+				d.uniqueIDMU.Lock()
+				if isBuilding := d.uniqueID[e.ID]; isBuilding {
+					// Delete the endpoint from the queue
+					// only if was marked as isBuilding
```

Comment:
```
// In a case where the same endpoint enters the
// building queue, while it was still being build,
// it will be marked as `false`/"not building",
// thus, we only delete the endpoint from the
// queue only if it is marked as isBuilding.
```
Better?